Completed the Test Automation homework task:

JustasGin opened the pull request:
- Created the calculator.spec.js file with the test cases for the [https://testsheepnz.github.io/BasicCalculator] webpage
 - Created the calculator.js file for the specific Calculator class to hold variables and methods that are used in calculator.spec.js
 - Created the calculator_notes.txt file that contains the analysis of the task

JustasGin committed:
Completed the Test Automation homework task:
 - Created the calculator.spec.js file with the test cases for the [https://testsheepnz.github.io/BasicCalculator] webpage
 - Created the calculator.js file for the specific Calculator class to hold variables and methods that are used in calculator.spec.js
 - Created the calculator_notes.txt file that contains the analysis of the task

Briedelis commented on cypress/integration/Homework/calculator_notes.txt:
```diff
@@ -0,0 +1,19 @@
+Dažniausiai pasikartojančios klaidos surastos:
```
Lithuanian comments? ;)

JustasGin replied:
My mistake, had a lot to say in Lithuanian ;)

Briedelis commented on cypress/integration/Homework/calculator.spec.js:
```diff
@@ -0,0 +1,408 @@
+            })
+            
+            it("TS6.5: Performs the operation 'Concatenate' with numbers " + firstNumber + " and " + secondNumber + " and clears the answer field", () => {
+                calculator.getFirstNumberField().clear()
```
Follow clean code principles and extract type in to a separate function.

Briedelis commented on cypress/integration/Homework/calculator.spec.js:
```diff
@@ -0,0 +1,408 @@
+        describe('TS6 (Build - ' + buildVersion + '): Calculator "Clear" button functionality testing with numbers', () => {
+            
+            before("Visits the initial page", () => {
+                calculator.visitInitialPage()
```
I see these two actions being repeated many types. Extract?

JustasGin replied:
I agree with this section. This could be extracted to a single point before all of the 'describe' sections. One question: should the website be visited before each test or only before each 'describe' section?

Briedelis replied:
It depends what you want to have - think about speed and reliability, one test failure should not affect other tests. Try out.

Briedelis commented on cypress/integration/Homework/calculator.spec.js:
```diff
@@ -0,0 +1,408 @@
+            })
+        })
+    })
+})
```
400+ lines. Whooooh. Please split.

Briedelis replied:
Or when you get rid of low value / repetitive tests it'll be fine.

JustasGin replied:
Removing the repetitive and unit tests would definitely resolve this issue. :)

Briedelis commented on cypress/page_objects/calculator.js:
```diff
@@ -0,0 +1,55 @@
+class Calculator {
+    
+    initialPage = "https://testsheepnz.github.io/BasicCalculator";
+    buildDropDownList = '#selectBuild';
```
It would be less cumbersome and more readable to have "buildVersion: () => cy.get('#selectBuild')". What do you think?

JustasGin replied:
I do agree with this solution. It would be easier to read and would save up some time as you have to search for the method that uses this value.

JustasGin replied:
Just a follow up, the correct syntax would be "buildVersion = () => cy.get('#selectBuild')", as type annotations are only supported in Typescript files. :)

Briedelis commented on cypress/integration/Homework/calculator.spec.js:
```diff
@@ -0,0 +1,408 @@
+
+// Build version names for the forEach cycle.
+var buildVersions = ["Prototype", "1", "2", "3", "4", "5", "6", "7", "8", "9"];
+//
```
Why there are these empty comments?

JustasGin replied:
I used these empty comments to highlight different sections. It is not necessary, just looked good to me. I will remove every empty comment section.

Briedelis commented on cypress/integration/Homework/calculator.spec.js:
```diff
@@ -0,0 +1,408 @@
+    [true, false].forEach((integersOnly) => {
+
+        // Operations for answer checking.
+        var addResultNumber = +firstNumber + +secondNumber;
```
"addResultNumber" name is not clear and does not reflect what the function does

Briedelis commented on cypress/integration/Homework/calculator.spec.js:
```diff
@@ -0,0 +1,408 @@
+            })
+        
+            it("TS1.1: Checks if the 'First number' field exists on the 'Add' operation and writes " + firstNumber + " into the field", () => {
+                calculator.getOperationDropDownList().select('Add')
```
It could be unit tests but not e2e. Simple test adding two numbers already covers this.

Briedelis replied:
Same with many tests below.

JustasGin replied:
I do agree with this. Got a little to carried away with writing these tests. I will remove them. :)

JustasGin replied:
Looking at it from the e2e perspective, the only viable tests are TS4, TS5 and TS6, am I correct? TS1 and TS2 sections only overlook one input field and TS3 evaluates only the 'Calculate' section.

Briedelis replied:
Didn't look through every test. Simple tests for sum, multiplication and similar should do the thing.

JustasGin committed:
Made changes based on the comments left in [https://github.com/MartynasLun/CypressWorkshop/pull/64]. Changed files:
 - calculator.spec.js
 - calculator_notes.txt
 - calculator.js

Briedelis commented on cypress/page_objects/calculator.js:
```diff
@@ -0,0 +1,24 @@
+
+    visitInitialPage = () => cy.visit("https://testsheepnz.github.io/BasicCalculator")
+
+    getBuildDropDownList = () => cy.get('#selectBuild')
```
Minor but I would prefer locator names without "get". Easier to read, faster to type :)

Briedelis commented on cypress/integration/Homework/calculator_notes.txt:
```diff
@@ -0,0 +1,18 @@
+firstSymbol  = C;  ; &$@;
+secondSymbol = P; P;   P;
+
+Considering all of the combined tests the only build that I was not able to break is number 5.
```
What about clear button functionality? ;)

Briedelis commented on cypress/integration/Homework/calculator.spec.js:
```diff
@@ -0,0 +1,263 @@
+            beforeSectionForPageAndIntegers(integersOnly, true, buildVersion);
+        
+            it.only(formItSentence("1", "1", "Add", "numbers", firstNumber, secondNumber, integersOnly, "and returns the answer " + addOperationResult), () => {
+                calculator.getFirstNumberField().clear()
```
I don't see clear functionality extracted as per my previous request. It is still being repeated in every test.

Briedelis replied:
For tests to be neat and clear I would expect something like this:
calculator.add(firstNumber, secondNumber)
calculator.answerField().should('have.value', expectedResult)

Clear functionality would be removed to beforeEach if needed
calculator.clear()
It might be not needed if you open the page beforeEach

Briedelis commented on cypress/integration/Homework/calculator.spec.js:
```diff
@@ -0,0 +1,263 @@
+                calculator.getSecondNumberField().type(secondNumber)
+                calculator.getOperationDropDownList().select('Add')
+                calculator.getCalculateButton().click()
+                calculator.getAnswerField().should('have.value', addOperationResult)
```
Does 'have.value' checks  exact match or that it consists the value?

JustasGin replied:
From the examples I have seen it checks if the value is an exact match. Am I missing something here?

Briedelis commented on cypress/integration/Homework/calculator.spec.js:
```diff
@@ -0,0 +1,263 @@
+    [true, false].forEach((integersOnly) => {
+
+        // Operations for answer checking.
+        var addOperationResult = +firstNumber + +secondNumber;
```
As discussed - we should not use these functions.

JustasGin committed:
Changes made to the files based on the comments in [https://github.com/MartynasLun/CypressWorkshop/pull/64].
Added an additional file numbers.js which contais an array of numbers and answers used in the automated tests.

Briedelis commented on cypress/integration/Homework/calculator.spec.js:
```diff
@@ -0,0 +1,194 @@
+
+var calculator = new Calculator
+
+// Build version names for the forEach cycle.
```
Variable name already tells what it is

Briedelis commented on cypress/integration/Homework/calculator.spec.js:
```diff
@@ -0,0 +1,194 @@
+// Build version names for the forEach cycle.
+var buildVersions = ["Prototype", "1", "2", "3", "4", "5", "6", "7", "8", "9"];
+
+// Variables for error messages.
```
Variable name already tells what it is

Briedelis commented on cypress/integration/Homework/calculator.spec.js:
```diff
@@ -0,0 +1,194 @@
+/// <reference types="cypress" />
+
+import Calculator from '../../page_objects/calculator.js'
+import numberAnswerArray from '../Homework/numbers.js'
```
File name is not clear - maybe testData would be better?

Briedelis commented on cypress/integration/Homework/calculator.spec.js:
```diff
@@ -0,0 +1,194 @@
+            beforeEachClearBothInputFields();
+
+            it.only(formItSentence("1", "1", "Add", "numbers", numberArray.firstNumber, numberArray.secondNumber, numberArray.integersOnly, "and returns the answer " + numberArray.addOperationResult), () => {
+                performCalculatorOperation(numberArray.firstNumber, numberArray.secondNumber, 'Add')
```
Nice!

Briedelis commented on cypress/integration/Homework/calculator.spec.js:
```diff
@@ -0,0 +1,194 @@
+            it.only(formItSentence("1", "4", "Divide", "numbers", numberArray.firstNumber, numberArray.secondNumber, numberArray.integersOnly, "and returns the answer " + numberArray.divideOperationResult) + " or gets the '" + divideByZeroError + "' error", () => {
+                performCalculatorOperation(numberArray.firstNumber, numberArray.secondNumber, 'Divide')
+
+                if (numberArray.secondNumber === 0 || +numberArray.secondNumber + 0 === 0) {
```
It shouts for a separate function.

Briedelis commented on cypress/integration/Homework/calculator.spec.js:
```diff
@@ -0,0 +1,194 @@
+            it(formItSentence("3", "4", "Divide", "numbers", numberArray.firstNumber, numberArray.secondNumber, numberArray.integersOnly, "and clears the answer field or gets the '" + divideByZeroError + "' error"), () => {
+                performCalculatorOperation(numberArray.firstNumber, numberArray.secondNumber, 'Divide')
+
+                if (numberArray.secondNumber === 0 || +numberArray.secondNumber + 0 === 0) {
```
Should go to a different function.

Briedelis commented on cypress/integration/Homework/calculator.spec.js:
```diff
@@ -0,0 +1,194 @@
+        describe.only(formDescribeSentence("3", buildVersion, "Calculator 'Clear' button functionality testing with numbers"), () => {
+
+            beforeSectionForPageAndIntegers(numberArray.integersOnly, false, buildVersion);
+            beforeEachClearBothInputFields();
```
It looks like one beforeEach with different methods being called inside would be cleaner than 2 different beforeEach.

Briedelis commented on cypress/integration/Homework/calculator.spec.js:
```diff
@@ -0,0 +1,194 @@
+    })
+}
+
+function beforeEachClearBothInputFields() {
```
I would have simple clear function and call it whatever needed - inside before each or it could be used somewhere else too.

Briedelis commented on cypress/integration/Homework/calculator.spec.js:
```diff
@@ -0,0 +1,194 @@
+    calculator.calculateButton().click()
+}
+
+function checkErrorField(errorMessage) {
```
As you are checking specifically that it is empty you should reflect it in the function name.

Briedelis commented on cypress/integration/Homework/calculator_notes.txt:
```diff
@@ -0,0 +1,17 @@
+is either to refresh the browser, or to change the build.
+I considered this fact as a 'feature' and wrote testing procedures accordingly.
+
+Similarities with [https://github.com/MartynasLun/CypressWorkshop/pull/72/files]:
```
Thanks, you can get rid of this section here.

Briedelis commented on cypress/integration/Homework/numbers.js:
```diff
@@ -0,0 +1,124 @@
+const numberAnswerArray = [
```
It looks like for the future it would be more readable to have a table with all test data. No need to change it now.

Briedelis replied:
You renamed the variable where you use this date but I would have the file name as test data or so too.

Briedelis reviewed (approved):
I would recommend to do those minor changes I commented on but these are not stoppers for a merge. 

Nice work!

JustasGin committed:
Made changes to the files based on the comments in [https://github.com/MartynasLun/CypressWorkshop/pull/64].

Briedelis commented on cypress/integration/Homework/calculator.spec.js:
```diff
@@ -0,0 +1,203 @@
+    }
+}
+
+function beforeSectionForPageAndIntegers(integersOnly, integerValidation, buildVersion) {
```
No need for this function. Use simple before above the tests. Easier to add more items later on.

Briedelis commented on cypress/integration/Homework/calculator.spec.js:
```diff
@@ -0,0 +1,203 @@
+    calculator.visitInitialPage()
+    calculator.buildDropDownList().select(buildVersion)
+
+    if (integersOnly && integerValidation) {
```
Multiple places you have this if - extract into function.

Briedelis commented on cypress/integration/Homework/calculator.spec.js:
```diff
@@ -0,0 +1,203 @@
+    calculator.secondNumberField().clear()
+}
+
+function beforeEachClearBothInputFields() {
```
I would get rid of this function at all. Simple beforeEach above the tests calling clear function would be enough as it would be easier to add more items in it.

JustasGin committed:
Made changes to the files based on the comments left in [https://github.com/MartynasLun/CypressWorkshop/pull/64].
